cmd/zed/zed_conf.c

```diff
@@ -32,38 +32,26 @@
 #include "zed_strings.h"
 
 /*
- * Return a new configuration with default values.
+ * Initialise the configuration with default values.
  */
-struct zed_conf *
-zed_conf_create(void)
+void
+zed_conf_init(struct zed_conf *zcp)
 {
-	struct zed_conf *zcp;
-
-	zcp = calloc(1, sizeof (*zcp));
-	if (!zcp)
-		goto nomem;
-
-	zcp->pid_fd = -1;
-	zcp->zedlets = NULL;		/* created via zed_conf_scan_dir() */
-	zcp->state_fd = -1;		/* opened via zed_conf_open_state() */
-	zcp->zfs_hdl = NULL;		/* opened via zed_event_init() */
-	zcp->zevent_fd = -1;		/* opened via zed_event_init() */
-	zcp->max_jobs = 16;
+	memset(zcp, 0, sizeof (*zcp));
 
-	if (!(zcp->pid_file = strdup(ZED_PID_FILE)))
-		goto nomem;
+	/* zcp->zfs_hdl opened in zed_event_init() */
+	/* zcp->zedlets created in zed_conf_scan_dir() */
 
-	if (!(zcp->zedlet_dir = strdup(ZED_ZEDLET_DIR)))
-		goto nomem;
+	zcp->pid_fd = -1;		/* opened in zed_conf_write_pid() */
+	zcp->state_fd = -1;		/* opened in zed_conf_open_state() */
+	zcp->zevent_fd = -1;		/* opened in zed_event_init() */
 
-	if (!(zcp->state_file = strdup(ZED_STATE_FILE)))
-		goto nomem;
-
-	return (zcp);
+	zcp->max_jobs = 16;
 
-nomem:
-	zed_log_die("Failed to create conf: %s", strerror(errno));
-	return (NULL);
+	if (!((zcp->pid_file = strdup(ZED_PID_FILE)) &&
+	    (zcp->zedlet_dir = strdup(ZED_ZEDLET_DIR)) &&
+	    (zcp->state_file = strdup(ZED_STATE_FILE))))
+		zed_log_die("Failed to create conf: %s", strerror(errno));
```

Comment:
This check isn't equivalent, we need to exit out if any of these allocations fail.

Reply:
It may be difficult to see due to the formatting, but I'm pretty sure this is right:
```c
if (!((zcp->pid_file = strdup(ZED_PID_FILE)) &&
    (zcp->zedlet_dir = strdup(ZED_ZEDLET_DIR)) &&
    (zcp->state_file = strdup(ZED_STATE_FILE))))

if (!((zcp->pid_file   = strdup(ZED_PID_FILE)  ) &&
      (zcp->zedlet_dir = strdup(ZED_ZEDLET_DIR)) &&
      (zcp->state_file = strdup(ZED_STATE_FILE))))

if (!( (zcp->pid_file   = strdup(ZED_PID_FILE)  )  &&
       (zcp->zedlet_dir = strdup(ZED_ZEDLET_DIR))  &&
       (zcp->state_file = strdup(ZED_STATE_FILE)) ))
```
which is by definition equivalent to
```c
if ( !(zcp->pid_file   = strdup(ZED_PID_FILE))   ||
     !(zcp->zedlet_dir = strdup(ZED_ZEDLET_DIR)) ||
     !(zcp->state_file = strdup(ZED_STATE_FILE)) )
```

But I found the `!(pf && zd && sf)` spelling simpler than the De Morgan.

Reply:
Ahh, you're right.  I was _repeatedly_ overlooking a layer of parenthesis.  Personally, I think writing it like this

```
if  !(zcp->pid_file = strdup(ZED_PID_FILE)) ||
    !(zcp->zedlet_dir = strdup(ZED_ZEDLET_DIR)) ||
    !(zcp->state_file = strdup(ZED_STATE_FILE)))
```

would be more readable at a glance but I don't have any objection to the `!(pf && zd && sf)` form.

Reply:
Reading it again, out of context now, and I tend to agree. Updated.